File: src/data/animationSockets.ts
import type {
  AnimationSetDef,
  SocketAtlasFingerprint,
  SocketDef,
  SocketFramePose,
  SocketSetDef,
} from './types';

/**
 * 动画挂点（sockets）的解析与失效判定。
 *
 * 数据住在动画包目录的 **sidecar** `sockets.json`，不进 anim.json：
 * anim.json 是产线产物（`export_gamedraft_anim*` 从零拼 dict），挂点是人工逐帧标的，
 * 两者生命周期不同。更要紧的是——挂点按**图集槽位**索引，重导出后槽位会漂移，
 * 所以必须能判"这份标注还对不对得上现在这张图集"，而不是让它悄悄错位。
 */

export const SOCKETS_SCHEMA_VERSION = 1;

/** `<dir>/anim.json` → `<dir>/sockets.json`；给不出目录时返回空串。 */
export function socketsJsonUrlForAnim(animJsonUrl: string): string {
  const url = (animJsonUrl || '').trim();
  const cut = url.lastIndexOf('/');
  if (cut < 0) return '';
  return `${url.slice(0, cut)}/sockets.json`;
}

function finiteNum(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null;
}

function parsePose(raw: unknown): SocketFramePose | null {
  if (typeof raw !== 'object' || raw === null) return null;
  const m = raw as Record<string, unknown>;
  const x = finiteNum(m.x);
  const y = finiteNum(m.y);
  if (x === null || y === null) return null;
  const pose: SocketFramePose = { x, y };
  const angle = finiteNum(m.angle);
  if (angle !== null && angle !== 0) pose.angle = angle;
  if (m.front === true) pose.front = true;
  const frame = finiteNum(m.frame);
  if (frame !== null) pose.frame = Math.trunc(frame);
  return pose;
}

/**
 * 从 anim 定义取出图集指纹（与 sockets.json 里存的那份比对）。
 *
 * **只取 cols / rows / slotCount**——这三个决定"槽位索引指向哪一格"。
 * 刻意**不含** cellWidth/cellHeight：格子像素尺寸变了而网格没变，意味着图集只是换了
 * 分辨率、槽位次序没动，格内归一化坐标依然成立，不该误杀。而且这两个值在 TS 侧经
 * `normalizeAnimationSetDef` 由纹理推导补全、编辑器侧读的是原始 JSON，放进指纹会让
 * 两侧对同一个包算出不同答案——8 个没写 cellWidth 的包会因此永远判失效。
 */
export function fingerprintOfAnim(def: AnimationSetDef): SocketAtlasFingerprint {
  return {
    cols: def.cols,
    rows: def.rows,
    slotCount: def.atlasFrames?.length ?? 0,
  };
}

/** 两份指纹是否一致（全等比较；任一维不同即视为图集换过了）。 */
export function fingerprintMatches(
  a: SocketAtlasFingerprint | null | undefined,
  b: SocketAtlasFingerprint | null | undefined,
): boolean {
  if (!a || !b) return false;
  return a.cols === b.cols && a.rows === b.rows && a.slotCount === b.slotCount;
}

/**
 * 解析 sockets.json 原始 JSON。结构性坏数据一律返回 null（当作没有挂点），
 * 单个坏 pose 跳过——挂点是表现层增益，任何时候都不该把角色本身弄挂。
 */
export function parseSocketSet(raw: unknown): SocketSetDef | null {
  if (typeof raw !== 'object' || raw === null) return null;
  const m = raw as Record<string, unknown>;
  const atlasRaw = m.atlas;
  if (typeof atlasRaw !== 'object' || atlasRaw === null) return null;
  const a = atlasRaw as Record<string, unknown>;
  const cols = finiteNum(a.cols);
  const rows = finiteNum(a.rows);
  const slotCount = finiteNum(a.slotCount);
  if (cols === null || rows === null || slotCount === null) return null;
  // `sockets` 可以缺省（一份只标了落脚帧、一个挂点都没有的 sidecar 是合法的）；
  // 但写了却不是对象就是结构坏了。
  const socketsRaw = m.sockets === undefined ? {} : m.sockets;
  if (typeof socketsRaw !== 'object' || socketsRaw === null) return null;

  const sockets: Record<string, SocketDef> = {};
  for (const [name, defRaw] of Object.entries(socketsRaw as Record<string, unknown>)) {
    if (typeof defRaw !== 'object' || defRaw === null) continue;
    const d = defRaw as Record<string, unknown>;
    const posesRaw = d.poses;
    if (typeof posesRaw !== 'object' || posesRaw === null) continue;
    const poses: Record<string, SocketFramePose> = {};
    for (const [slot, poseRaw] of Object.entries(posesRaw as Record<string, unknown>)) {
      const pose = parsePose(poseRaw);
      if (pose) poses[slot] = pose;
    }
    const out: SocketDef = { poses };
    if (typeof d.label === 'string' && d.label.trim()) out.label = d.label.trim();
    sockets[name] = out;
  }

  const version = finiteNum(m.schemaVersion) ?? SOCKETS_SCHEMA_VERSION;
  return {
    schemaVersion: Math.trunc(version),
    atlas: { cols, rows, slotCount },
    sockets,
    contactSlots: parseContactSlots(m.contactSlots, slotCount),
  };
}

/**
 * 落脚帧槽位：只收 `0 <= 整数 < slotCount`，去重升序；不是数组/整份坏掉 ⇒ 空（无脚步）。
 * 单个坏项跳过——与 pose 同一口径：标注是表现层增益，任何时候不该把角色本身弄挂。
 */
function parseContactSlots(raw: unknown, slotCount: number): number[] {
  if (!Array.isArray(raw)) return [];
  const out = new Set<number>();
  for (const v of raw) {
    if (typeof v !== 'number' || !Number.isInteger(v) || v < 0) continue;
    if (slotCount > 0 && v >= slotCount) continue;
    out.add(v);
  }
  return Array.from(out).sort((a, b) => a - b);
}

/** 解算挂点局部位姿要的宿主状态（SpriteEntity 与编辑器画布共用同一套输入）。 */
export interface SocketHostFrame {
  /** 精灵在世界中的宽/高（anim.json 归一化后的值） */
  worldWidth: number;
  worldHeight: number;
  /** 透视系数（近大远小）；1 = 不缩 */
  depthScale: number;
  /** 朝向：1 右、-1 左（镜像） */
  facing: 1 | -1;
  /** 跳跃视觉抬升（容器局部 px，负=向上）；非跳跃时 0 */
  visualLiftY: number;
  /**
   * 宿主精灵的锚点（世界包围盒内归一化，x 0=左 1=右、y 0=顶 1=底）。
   * **缺省 (0.5, 1) = 底中 = 脚底**，即锚点可配之前写死的那个值。
   *
   * 为什么必须知道：挂点标注是**格内**归一化（`raw.x/raw.y` 同样 0..1），而返回的是
   * **容器局部**坐标，两者之间差的正是"格子的哪一点压在容器原点上"= 锚点。
   * 不传就按脚底算，宿主一旦把锚点挪到圆心，刀就会整体错半个身位（而且不报错）。
   */
  anchorX?: number;
  anchorY?: number;
}

/** 解算出的容器局部位姿 */
export interface SocketLocalPose {
  x: number;
  y: number;
  angleDeg: number;
  front: boolean;
  frame: number | null;
  scale: number;
  facing: 1 | -1;
}

/**
 * 把格内归一化标注解算成**容器局部**位姿。运行时与编辑器画布共用这一处，
 * 免得"编辑器里对齐了、游戏里差半个身位"。
 *
 * 换算链（与 `SpriteEntity.applySpriteScale` / `getAuthoredBubbleAnchorLocalY` 同源）：
 * - `x`：格内 0..1 → 以底中锚为原点 → 乘世界宽与透视 → 乘朝向（镜像整体翻到另一侧）
 * - `y`：格内 0=顶 1=底(脚线) → 脚点为 0、向上为负 → 加跳跃视觉抬升
 * - `angle`：朝左时取反（镜像后顺时针变逆时针）
 */
export function socketPoseToLocal(
  raw: SocketFramePose,
  host: SocketHostFrame,
): SocketLocalPose {
  const d = host.depthScale > 0 && Number.isFinite(host.depthScale) ? host.depthScale : 1;
  const sign = host.facing;
  // 格内归一化 → 容器局部：减掉的正是宿主锚点（缺省底中，与改造前逐位相同）
  const ax = typeof host.anchorX === 'number' && Number.isFinite(host.anchorX) ? host.anchorX : 0.5;
  const ay = typeof host.anchorY === 'number' && Number.isFinite(host.anchorY) ? host.anchorY : 1;
  return {
    x: (raw.x - ax) * host.worldWidth * d * sign,
    y: host.visualLiftY + (raw.y - ay) * host.worldHeight * d,
    angleDeg: (raw.angle ?? 0) * sign,
    front: raw.front === true,
    frame: typeof raw.frame === 'number' ? raw.frame : null,
    scale: d,
    facing: sign,
  };
}

/** 解析结果 + 是否与当前图集对得上（对不上时 `set` 仍返回，供编辑器提示重标）。 */
export interface ResolvedSockets {
  set: SocketSetDef | null;
  /** true = 指纹对不上，运行时必须当作没有挂点 */
  stale: boolean;
}

/**
 * 把原始 sidecar JSON 解析成运行时可用的挂点集。
 * 指纹对不上时 `stale=true`：运行时据此拒绝使用（宁可不挂，不挂错）。
 */
export function resolveSockets(raw: unknown, animDef: AnimationSetDef): ResolvedSockets {
  const set = parseSocketSet(raw);
  if (!set) return { set: null, stale: false };
  const stale = !fingerprintMatches(set.atlas, fingerprintOfAnim(animDef));
  return { set, stale };
}

/** AssetManager 里本模块用到的那一小片（避免 data 层反向依赖 core 的具体类）。 */
export interface OptionalJsonLoader {
  loadOptionalJson<T = unknown>(path: string): Promise<T | null>;
}

/**
 * 随动画包载它的挂点 sidecar。**没有 sockets.json 是绝大多数包的常态**，
 * 返回 `{set:null,stale:false}`（等于没挂点），不报错、不打日志。
 */
export async function loadSocketsForAnim(
  loader: OptionalJsonLoader,
  animJsonUrl: string,
  animDef: AnimationSetDef,
): Promise<ResolvedSockets> {
  const url = socketsJsonUrlForAnim(animJsonUrl);
  if (!url) return { set: null, stale: false };
  const raw = await loader.loadOptionalJson(url);
  if (raw === null) return { set: null, stale: false };
  return resolveSockets(raw, animDef);
}

File: tools/editor/shared/animation_sockets.py
"""动画挂点（sockets）与落脚帧（contactSlots）的编辑器侧读写与几何。

数据是 sidecar：``<动画包目录>/sockets.json``，**产线永不触碰**（导出器从零拼 anim.json，
挂点跟它生命周期不同）。按**图集槽位**索引——同一张图＝同一个手的位置＝同一只脚落地，
导出的去重合并还强化了这一点。落脚帧（``contactSlots``）与挂点同住一份文件、同一份指纹，
在同一个面板里看着图逐帧标；运行时据它决定哪一帧播脚步声。

``atlas`` 是图集指纹：重导出后槽位会漂移，对不上就整份判失效（stale），拒绝使用并提示重标。
盲用漂移后的槽位号会静默挂错位置，比不挂更坏。

⚠ 只是网格指纹（cols/rows/槽位数）：重抠图但网格没变**检测不出来**，
那种情况得靠人重看一遍。这是本机制已知且刻意的边界。

`socket_pose_to_local` 是 ``src/data/animationSockets.ts`` 的 `socketPoseToLocal` 的
**跨语言镜像**——两边必须逐值一致，由 `test_socket_pose_parity.py` 钉死
（norms 第 8 条：手工镜像必配语义级 parity）。
"""
from __future__ import annotations

import copy
from pathlib import Path
from typing import Any

from ..file_io import read_json, write_json

SOCKETS_SCHEMA_VERSION = 1
SOCKETS_FILENAME = "sockets.json"

#: 一条 pose 允许出现的键（写盘时按此顺序，缺省值不落键）
POSE_KEYS = ("x", "y", "angle", "front", "frame")

#: 落脚帧：脚触地的图集槽位（升序去重）。与挂点同住 sidecar、同一份指纹——
#: 都是"看着这一格画的是什么"逐帧标出来的，重导出槽位漂移时一起判失效。
#: 运行时 `SpriteEntity.isContactFrameAt` 按它决定哪一帧播脚步声；空 = 这个包没有脚步。
CONTACT_SLOTS_KEY = "contactSlots"


def sockets_path_for_bundle(animation_bundles_path: Path, bundle_id: str) -> Path:
    """``<animation>/<bundle>/sockets.json``。"""
    return Path(animation_bundles_path) / str(bundle_id).strip() / SOCKETS_FILENAME


def fingerprint_of_anim(anim: dict) -> dict[str, Any]:
    """从 anim.json 取图集指纹（与 sockets.json 里存的那份比对）。

    **只取 cols / rows / slotCount**（与 TS 侧 `fingerprintOfAnim` 逐字同口径）。
    刻意不含 cellWidth/cellHeight：TS 侧那两个值经 normalizeAnimationSetDef 由纹理推导
    补全、这里读的是原始 JSON，放进指纹会让两侧对同一个包算出不同答案。
    """
    def _num(v: object, default: float = 0.0) -> float:
        return float(v) if isinstance(v, (int, float)) and not isinstance(v, bool) else default

    frames = anim.get("atlasFrames")
    return {
        "cols": int(_num(anim.get("cols"))),
        "rows": int(_num(anim.get("rows"))),
        "slotCount": len(frames) if isinstance(frames, list) else 0,
    }


def fingerprint_matches(a: dict | None, b: dict | None) -> bool:
    if not isinstance(a, dict) or not isinstance(b, dict):
        return False
    keys = ("cols", "rows", "slotCount")
    return all(a.get(k) == b.get(k) for k in keys)


def empty_socket_set(anim: dict) -> dict[str, Any]:
    """按当前图集指纹起一份空挂点集。"""
    return {
        "schemaVersion": SOCKETS_SCHEMA_VERSION,
        "atlas": fingerprint_of_anim(anim),
        "sockets": {},
    }


def load_socket_set(path: Path) -> dict[str, Any] | None:
    """读 sockets.json；不存在或坏掉返回 None（当作没有挂点）。"""
    p = Path(path)
    if not p.is_file():
        return None
    try:
        data = read_json(p)
    except (OSError, ValueError):
        return None
    return data if isinstance(data, dict) else None


def save_socket_set(path: Path, data: dict) -> Path:
    """写 sockets.json（走 write_json：UTF-8 / 2 空格 / 中文不转义 / 保留键序 / 末尾换行）。

    **整份为空时删文件**：既没有挂点、也没有落脚帧，就不该在包目录里留一个空壳，
    否则每个动画包都多一个永远是 `{}` 的文件。只标了落脚帧、一个挂点都没有的
    sidecar 是合法且常见的（绝大多数会走路的包都这样），必须保留。
    """
    p = Path(path)
    sockets = data.get("sockets")
    has_sockets = isinstance(sockets, dict) and bool(sockets)
    if not has_sockets and not contact_slots_of(data):
        if p.is_file():
            p.unlink()
        return p
    p.parent.mkdir(parents=True, exist_ok=True)
    write_json(p, data)
    return p


def normalize_contact_slots(raw: object, slot_count: int = 0) -> list[int]:
    """落脚帧槽位规整：只收 `0 <= 整数 < slot_count`（slot_count<=0 时不查上界），去重升序。

    与 TS 侧 `parseContactSlots` 同口径：bool 不算整数（`True` 落成 JSON 是 `true`，
    运行时 `Number.isInteger(true)` 为 false），坏项跳过而不是整份作废。
    """
    if not isinstance(raw, list):
        return []
    out: set[int] = set()
    for v in raw:
        if isinstance(v, bool) or not isinstance(v, (int, float)):
            continue
        if isinstance(v, float) and not float(v).is_integer():
            continue
        s = int(v)
        if s < 0 or (slot_count > 0 and s >= slot_count):
            continue
        out.add(s)
    return sorted(out)


def contact_slots_of(data: dict | None) -> list[int]:
    """读一份（内存态或磁盘态）挂点集里的落脚帧槽位；缺省/坏形状 = 空。"""
    if not isinstance(data, dict):
        return []
    return normalize_contact_slots(data.get(CONTACT_SLOTS_KEY))


def set_contact_slot(data: dict, slot: int, on: bool) -> bool:
    """把某个图集槽位标成/取消落脚帧（就地改 data）。返回是否真的变了。"""
    cur = set(contact_slots_of(data))
    s = int(slot)
    if on:
        if s in cur:
            return False
        cur.add(s)
    else:
        if s not in cur:
            return False
        cur.discard(s)
    if cur:
        data[CONTACT_SLOTS_KEY] = sorted(cur)
    else:
        data.pop(CONTACT_SLOTS_KEY, None)
    return True


def normalize_pose(raw: dict) -> dict[str, Any] | None:
    """规整一条 pose：x/y 必需，其余缺省值不落键（往返干净）。"""
    if not isinstance(raw, dict):
        return None
    def _num(key: str):
        v = raw.get(key)
        return float(v) if isinstance(v, (int, float)) and not isinstance(v, bool) else None
    x, y = _num("x"), _num("y")
    if x is None or y is None:
        return None
    out: dict[str, Any] = {"x": round(x, 5), "y": round(y, 5)}
    angle = _num("angle")
    if angle is not None and round(angle, 3) != 0:
        out["angle"] = round(angle, 3)
    if raw.get("front") is True:
        out["front"] = True
    frame = _num("frame")
    if frame is not None:
        out["frame"] = int(frame)
    return out


def sanitize_socket_set(data: dict, anim: dict) -> dict[str, Any]:
    """把编辑器内存态整理成可写盘的形状（指纹按当前图集刷新、坏 pose 丢弃）。"""
    out: dict[str, Any] = {
        "schemaVersion": SOCKETS_SCHEMA_VERSION,
        "atlas": fingerprint_of_anim(anim),
        "sockets": {},
    }
    # 落脚帧：只在非空时落键（没标过的包不该多出一个 `[]`，往返干净）；
    # 指纹已按当前图集刷新，所以槽位上界也按当前图集裁——超出的必然是漂移后的垃圾。
    contact = normalize_contact_slots(
        data.get(CONTACT_SLOTS_KEY) if isinstance(data, dict) else None,
        int(out["atlas"].get("slotCount") or 0),
    )
    if contact:
        out[CONTACT_SLOTS_KEY] = contact
    raw_sockets = data.get("sockets") if isinstance(data, dict) else None
    if not isinstance(raw_sockets, dict):
        return out
    for name, sock in raw_sockets.items():
        if not isinstance(sock, dict):
            continue
        poses_raw = sock.get("poses")
        if not isinstance(poses_raw, dict):
            continue
        poses: dict[str, Any] = {}
        for slot, pose in sorted(poses_raw.items(), key=lambda kv: int(kv[0]) if str(kv[0]).isdigit() else 1 << 30):
            norm = normalize_pose(pose)
            if norm is not None:
                poses[str(slot)] = norm
        if not poses:
            continue
        entry: dict[str, Any] = {}
        label = sock.get("label")
        if isinstance(label, str) and label.strip():
            entry["label"] = label.strip()
        entry["poses"] = poses
        out["sockets"][str(name)] = entry
    return out


def socket_pose_to_local(
    pose: dict,
    *,
    world_width: float,
    world_height: float,
    depth_scale: float = 1.0,
    facing: int = 1,
    visual_lift_y: float = 0.0,
) -> dict[str, Any]:
    """把格内归一化标注解算成**容器局部**位姿。

    ⚠ 这是 ``src/data/animationSockets.ts::socketPoseToLocal`` 的跨语言镜像，
    改一处必改两处（`test_socket_pose_parity.py` 逐值对账）。
    """
    d = depth_scale if isinstance(depth_scale, (int, float)) and depth_scale > 0 else 1.0
    try:
        d = float(d)
        if d != d or d in (float("inf"), float("-inf")):  # NaN / inf
            d = 1.0
    except (TypeError, ValueError):
        d = 1.0
    sign = -1 if facing < 0 else 1
    x = float(pose.get("x", 0.0))
    y = float(pose.get("y", 0.0))
    angle = float(pose.get("angle", 0.0) or 0.0)
    frame = pose.get("frame")
    return {
        "x": (x - 0.5) * world_width * d * sign,
        "y": visual_lift_y + (y - 1.0) * world_height * d,
        "angleDeg": angle * sign,
        "front": pose.get("front") is True,
        "frame": int(frame) if isinstance(frame, (int, float)) and not isinstance(frame, bool) else None,
        "scale": d,
        "facing": sign,
    }


def copy_pose_between_slots(data: dict, socket: str, src_slot: int, dst_slot: int) -> bool:
    """把某挂点上一帧的标注复制到另一帧（逐帧标注的主要省力手段）。"""
    sock = (data.get("sockets") or {}).get(socket)
    if not isinstance(sock, dict):
        return False
    poses = sock.get("poses")
    if not isinstance(poses, dict):
        return False
    src = poses.get(str(src_slot))
    if not isinstance(src, dict):
        return False
    poses[str(dst_slot)] = copy.deepcopy(src)
    return True


def interpolate_poses(data: dict, socket: str, slots: list[int]) -> int:
    """在给定槽位序列上做线性插值：两端已标注、中间空的按比例补。

    逐帧手点 3176 个槽位不现实，关键帧+插值是可行性前提而不是锦上添花。
    只补**空的**中间帧（已标的不覆盖），返回补了几帧。
    """
    sock = (data.get("sockets") or {}).get(socket)
    if not isinstance(sock, dict):
        return 0
    poses = sock.get("poses")
    if not isinstance(poses, dict):
        return 0
    keyed = [i for i, s in enumerate(slots) if isinstance(poses.get(str(s)), dict)]
    if len(keyed) < 2:
        return 0
    filled = 0
    for a, b in zip(keyed, keyed[1:]):
        if b - a < 2:
            continue
        pa = poses[str(slots[a])]
        pb = poses[str(slots[b])]
        for i in range(a + 1, b):
            if isinstance(poses.get(str(slots[i])), dict):
                continue
            t = (i - a) / (b - a)
            mid: dict[str, Any] = {
                "x": round(float(pa["x"]) + (float(pb["x"]) - float(pa["x"])) * t, 5),
                "y": round(float(pa["y"]) + (float(pb["y"]) - float(pa["y"])) * t, 5),
            }
            aa = float(pa.get("angle", 0) or 0)
            ab = float(pb.get("angle", 0) or 0)
            if round(aa, 3) or round(ab, 3):
                mid["angle"] = round(aa + (ab - aa) * t, 3)
            # front / frame 是离散量，插值无意义：跟起点走
            if pa.get("front") is True:
                mid["front"] = True
            if isinstance(pa.get("frame"), int):
                mid["frame"] = pa["frame"]
            poses[str(slots[i])] = mid
            filled += 1
    return filled
